pymc_experimental/inference/pathfinder/pathfinder.py

```diff
@@ -0,0 +1,980 @@
+
+from collections.abc import Callable
+from importlib.util import find_spec
+from typing import Literal
+
+import arviz as az
+import blackjax
+import filelock
+import jax
+import numpy as np
+import pymc as pm
+import pytensor
+import pytensor.tensor as pt
+
+from packaging import version
+from pymc import Model
+from pymc.backends.arviz import coords_and_dims_for_inferencedata
+from pymc.blocking import DictToArrayBijection, RaveledVars
+from pymc.initial_point import make_initial_point_fn
+from pymc.model import modelcontext
+from pymc.model.core import Point
+from pymc.pytensorf import compile_pymc
+from pymc.sampling.jax import get_jaxified_graph
+from pymc.util import (
+    CustomProgress,
+    RandomSeed,
+    _get_seeds_per_chain,
+    default_progress_theme,
+    get_default_varnames,
+)
+from pytensor.compile.io import In
+from pytensor.compile.mode import FAST_COMPILE
+from pytensor.graph import Apply, Op, vectorize_graph
+from rich.console import Console
+
+from pymc_experimental.inference.pathfinder.importance_sampling import (
+    importance_sampling as _importance_sampling,
+)
+from pymc_experimental.inference.pathfinder.lbfgs import LBFGSInitFailed, LBFGSOp
+
+logger = logging.getLogger(__name__)
+
+REGULARISATION_TERM = 1e-8
+
+
+def get_jaxified_logp_of_ravel_inputs(
+    model: Model,
+) -> Callable:
+    """
+    Get jaxified logp function and ravel inputs for a PyMC model.
+
+    Parameters
+    ----------
+    model : Model
+        PyMC model to jaxify.
+
+    Returns
+    -------
+    tuple[Callable, DictToArrayBijection]
+        A tuple containing the jaxified logp function and the DictToArrayBijection.
+    """
+
+    # TODO: set jacobian = True to avoid very high values for pareto k.
+
+    new_logprob, new_input = pm.pytensorf.join_nonshared_inputs(
+        model.initial_point(), (model.logp(),), model.value_vars, ()
+    )
+
+    logp_func_list = get_jaxified_graph([new_input], new_logprob)
+
+    def logp_func(x):
+        return logp_func_list(x)[0]
+
+    return logp_func
+
+
+def get_logp_dlogp_of_ravel_inputs(model: Model, jacobian: bool = True):
+    # setting jacobian = True, otherwise get very high values for pareto k.
+    outputs, inputs = pm.pytensorf.join_nonshared_inputs(
+        model.initial_point(),
+        [model.logp(jacobian=jacobian), model.dlogp(jacobian=jacobian)],
+        model.value_vars,
+    )
+
+    logp_func = compile_pymc(
+        [inputs], outputs[0], mode=pytensor.compile.mode.Mode(linker="cvm_nogc")
+    )
+    logp_func.trust_input = True
+
+    dlogp_func = compile_pymc(
+        [inputs], outputs[1], mode=pytensor.compile.mode.Mode(linker="cvm_nogc")
+    )
+    dlogp_func.trust_input = True
+
+    return logp_func, dlogp_func
+
+
+def convert_flat_trace_to_idata(
+    samples,
+    include_transformed=False,
+    postprocessing_backend="cpu",
+    inference_backend="pymc",
+    model=None,
+    importance_sampling: Literal["psis", "psir", "identity", "none"] = "psis",
```

Comment:
The `importance_sampling` argument isn't passed anywhere and is only relevant when it is "none". Is this intended?